allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,244 @@
+
+from allennlp.common import Params
+from allennlp.data import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules.token_embedders import Embedding
+from allennlp.modules import FeedForward
+from allennlp.modules import Seq2SeqEncoder, TimeDistributed, TextFieldEmbedder
+from allennlp.nn import util, InitializerApplicator, RegularizerApplicator
+from allennlp.training.metrics import MentionRecall, ConllScores
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+@Model.register("coref")
+class CoreferenceResolver(Model):
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 context_layer: Seq2SeqEncoder,
+                 mention_feedforward: FeedForward,
+                 antecedent_feedforward: FeedForward,
+                 feature_size: int,
+                 max_span_width: int,
+                 spans_per_word: float,
+                 max_antecedents: int,
+                 lexical_dropout: float = 0.2,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+        super(CoreferenceResolver, self).__init__(vocab, regularizer)
+        self._text_field_embedder = text_field_embedder
+        self._context_layer = context_layer
+        self._mention_feedforward = TimeDistributed(mention_feedforward)
+        self._antecedent_feedforward = TimeDistributed(antecedent_feedforward)
+        self._mention_scorer = TimeDistributed(torch.nn.Linear(mention_feedforward.get_output_dim(), 1))
+        self._antecedent_scorer = TimeDistributed(torch.nn.Linear(antecedent_feedforward.get_output_dim(), 1))
+        self._head_scorer = TimeDistributed(torch.nn.Linear(context_layer.get_output_dim(), 1))
+        self._distance_embedding = Embedding(10, feature_size) # 10 possible distance buckets.
+        self._span_width_embedding = Embedding(max_span_width, feature_size)
+        self._max_span_width = max_span_width
+        self._spans_per_word = spans_per_word
+        self._max_antecedents = max_antecedents
+        self._mention_recall = MentionRecall()
+        self._conll_scores = ConllScores()
+        if lexical_dropout > 0:
+            self._lexical_dropout = torch.nn.Dropout(p=lexical_dropout)
+        else:
+            self._lexical_dropout = lambda x: x
+        initializer(self)
+
+    @staticmethod
+    def _bucket_distance(distances):
+        # Buckets: [0, 1, 2, 3, 4, 5-7, 8-15, 16-31, 32-63, 64+].
+        logspace_idx = (distances.float().log()/math.log(2)).floor().long() + 3
+        use_identity = (distances <= 4).long()
+        combined_idx = use_identity * distances + (1 + (-1 * use_identity)) * logspace_idx
+        return combined_idx.clamp(0, 9)
+
+    def forward(self,  # type: ignore
+                text: Dict[str, torch.LongTensor],
+                span_starts: torch.IntTensor,
```

Comment:
Looking at your loss function, too, it doesn't look like span start and end are part of the loss.  So they are just considered input?  In a demo, we'd have to run some mention detector on the text first, before inputting to this?  Or am I missing something?

Reply:
It's jointly inferring and learning mention detection and coref, which is why they are inseparable. The span starts and span ends are implicitly part of the loss, since the span representations depend on them.

Reply:
Ok, the confusion was due to me not looking closely enough at how `span_starts` and `span_ends` are created. You just enumerate all possible spans. It's just the labels that should be optional, then.

Reply:
Fixed.